pkg/chunk/aws_storage_client.go

```diff
@@ -331,28 +340,232 @@ func (a dynamoDBRequestAdapter) HasNextPage() bool {
+	}
+	return result, nil
+}
+
+func (a awsStorageClient) getS3Chunk(ctx context.Context, chunk Chunk) (Chunk, error) {
 	var resp *s3.GetObjectOutput
 	err := instrument.TimeRequestHistogram(ctx, "S3.GetObject", s3RequestDuration, func(ctx context.Context) error {
 		var err error
 		resp, err = a.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
 			Bucket: aws.String(a.bucketName),
-			Key:    aws.String(key),
+			Key:    aws.String(chunk.externalKey()),
 		})
 		return err
 	})
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
 	}
 	defer resp.Body.Close()
 	buf, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
+	}
+	if err := chunk.decode(buf); err != nil {
+		return Chunk{}, err
 	}
-	return buf, nil
+	return chunk, nil
 }
 
-func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+func (a awsStorageClient) getDynamoDBChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	outstanding := dynamoDBReadRequest{}
+	chunksByKey := map[string]Chunk{}
+	for _, chunk := range chunks {
+		key := chunk.externalKey()
+		chunksByKey[key] = chunk
+		tableName := a.chunkTableFor(chunk.From)
+		outstanding.Add(tableName, key, nil)
+	}
+
+	result := []Chunk{}
+	unprocessed := dynamoDBReadRequest{}
+	backoff, numRetries := minBackoff, 0
+	for outstanding.Len()+unprocessed.Len() > 0 {
+		requests := dynamoDBReadRequest{}
+		requests.takeReqs(unprocessed, dynamoDBMaxReadBatchSize)
+		requests.takeReqs(outstanding, dynamoDBMaxReadBatchSize)
+
+		var response *dynamodb.BatchGetItemOutput
+		err := instrument.TimeRequestHistogram(ctx, "DynamoDB.BatchGetItemPages", dynamoRequestDuration, func(ctx context.Context) error {
+			var err error
+			response, err = a.DynamoDB.BatchGetItemWithContext(ctx, &dynamodb.BatchGetItemInput{
+				RequestItems:           outstanding,
+				ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+			})
+			return err
+		})
+
+		for _, cc := range response.ConsumedCapacity {
+			dynamoConsumedCapacity.WithLabelValues("DynamoDB.BatchGetItemPages").
+				Add(float64(*cc.CapacityUnits))
+		}
+
+		if err != nil {
+			for tableName := range outstanding {
+				recordDynamoError(tableName, err)
+			}
+		}
+
+		for _, items := range response.Responses {
+			for _, item := range items {
+				key, ok := item[hashKey]
+				if !ok || key == nil || key.S == nil {
+					return result, fmt.Errorf("Got response from DynamoDB with no hash key: %+v", item)
+				}
+
+				chunk, ok := chunksByKey[*key.S]
+				if !ok {
+					return result, fmt.Errorf("Got response from DynamoDB with for chunk I didn't ask for: %s", *key.S)
+				}
+
+				buf, ok := item[valueKey]
+				if !ok || buf == nil || buf.B == nil {
+					return result, fmt.Errorf("Got response from DynamoDB with no value: %+v", item)
+				}
+
+				if err := chunk.decode(buf.B); err != nil {
+					return result, err
+				}
+
+				result = append(result, chunk)
+			}
+		}
```

Comment:
I'd extract this logic (that converts responses to chunks)  into a separate function. AFAICT, it doesn't affect the looping/tracking logic that's elsewhere, so it'd make sense to isolate it.